### PR TITLE
Running containers with kubedock (crw-3367)

### LINKED ISSUE #22817: Docs: how to configure a cluster and a workspace to execute podman run
### Is your enhancement related to a problem? Please describe

there is currently no Eclipse Che / Dev Spaces documentation about Kubedock and running Podman from within a workspace

### Describe the solution you'd like

create a doc about configuring a cluster and a workspace to execute podman run

### Describe alternatives you've considered

_No response_

### Additional context

https://issues.redhat.com/browse/CRW-5731
https://issues.redhat.com/browse/CRW-3367
https://github.com/eclipse-che/che-docs/pull/2710

### PR DESCRIPTION
**Pull request by deerskindoll**
## Pull Request checklist

The author and the reviewers validate the content of this pull request with the following checklist, in addition to the [automated tests](code_review_checklist.adoc).

- Any procedure:
  - [ ] Successfully tested.
- Any page or link rename:
  - [ ] The page contains a redirection for the previous URL.
  - Propagate the URL change in:
    - [ ] Dashboard [default branding data](https://github.com/eclipse-che/che-dashboard/blob/main/packages/dashboard-frontend/src/services/bootstrap/branding.constant.ts)
- [ ] Builds on [Eclipse Che hosted by Red Hat](https://workspaces.openshift.com).
- [ ] the *`Validate language on files added or modified`* step reports no vale warnings.

### CODE REVIEW

**Comment by deerskindoll** [comment]
feedback from  Anton:

I would really appreciate if we could add some more context explaining why kubedock is useful. The docs says
Use kubedock to run containers as Kubernetes pods. 

And while that is technically true, you can also use devfile to run containers. Or k8s deployment. Kubedock is minimal container engine implementation which gives you podman(or docker)-like experience inside Dev Spaces Workspace. Developers are used to using podman/docker during their local development workflow, and kubedock brings that experience onto the OpenShift, inside Dev Spaces.. This is otherwise incredibly hard to achieve due to the inherent OpenShift security restrictions. Kubedock is specifically useful when dealing with ad-hoc, ephemeral and testing containers. Suitable use cases are  (list is not exhaustive) 

1) Executing applications tests which rely on Testcontainers framework - kubedock will dynamically spin up required test container image , on the fly, as an OpenShift pod, giving user seamless experience. 

2) Quarkus Dev Services - which are internally relying on Testcontainers as well. 

3) When you are in need to you to run a container stored in remote container registry, for local development purposes and for whatever reasons you don't want to pollute the devfile . It's very important to mention (even though this is obvious!) that even though it gives you podman-like experience which you are used to from your local development environment, the absolute hard requirement is that the particular image which is being used via kubedock must be compatible with OpenShift. I specifically have security requirements in mind - i.e. not every container image which can be executed in your local machine, can also be executed on OpenShift. So if you do podman run <image-incompatible-with-openshift>  - it will fail with kubedock, and it will work locally. (edited) 

Once the use cases/motivation is explained, it would really good if we can share an example devfile.yaml, but ideally with accompanied git repo as well. I know this is hard, because including public github link in official documentation is always risky, so not sure what is the general rule of thumb for tech writers.

Container Image Guidelines for OpenSHift are here https://docs.openshift.com/container-platform/4.15/openshift_images/create-images.html - For example, the kafka-native testcontainer wasn't compatible with openshift - it only worked in local environment. So user who used this testcontinaer locally, would be surprised if he tried to use it via kubedock in dev spaces. We had to reach out to upstream community and they had to refactor that image, so it would become OCP-compatible (luckily, the image owner was Red Hatter as well..). I am just sharing all my experience accumulated over the past weeks..

**Comment by deerskindoll** [comment]
@agiertli let me know if you like it. I didn't include some of the details because I felt the content is pretty self-explanatory.

File: modules/administration-guide/pages/running-containers-with-kubedock.adoc [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+====
+Using `podman` commands with kubedock has following limitations
+
+* The `podman build -t <image> . && podman run <image>` command will fail. Use `podman build -t <image> . && podman push <image> && podman run <image>` instead.
+* The `podman generate kube` command is not supported.
+* `--env` option causes the `podman run` command to fail.
+====
+
+.Prerequisites
+* An image compliant with Openshift security requirements.
```
Comment (agiertli) [inline modules/administration-guide/pages/running-containers-with-kubedock.adoc]: I would suggest to list those requirements explicitly

File: modules/administration-guide/pages/running-containers-with-kubedock.adoc [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+.Process
+. Add `KUBEDOCK_ENABLED=true` environment variable to the devfile.
```
Comment (agiertli) [inline modules/administration-guide/pages/running-containers-with-kubedock.adoc]: Please share an actual devfile example with such env variable defined, it makes it easier for users to follow this.

File: modules/administration-guide/pages/running-containers-with-kubedock.adoc [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+[NOTE]
+====
+You must configure podman or docker API to point to kubedock setting
+`CONTAINER_HOST=tcp://127.0.0.1:2475` or `DOCKER_HOST=tcp://127.0.0.1:2475` when running containers.
+However, you must configure podman to point to local podman when building the container.
```
Comment (agiertli) [inline modules/administration-guide/pages/running-containers-with-kubedock.adoc]: There must be an example devfile shared in documentation.

I would also mention that user can supply additional kubedock parameters via KUBEDOCK_PARAMS as demonstrated here:

https://github.com/agiertli/quarkus-workshops/blob/main/devfile.yaml#L14

File: modules/administration-guide/pages/running-containers-with-kubedock.adoc [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+[IMPORTANT]
+====
+You must configure the Podman or docker API to point to kubedock setting
+`CONTAINER_HOST=tcp://127.0.0.1:2475` or `DOCKER_HOST=tcp://127.0.0.1:2475` when running containers.
+However, you must configure Podman to point to local Podman when building the container.
+====
+
+.Additional resources
+* link:https://github.com/agiertli/quarkus-workshops/blob/main/devfile.yaml[Sample devfile]
```
Comment (agiertli) [inline modules/administration-guide/pages/running-containers-with-kubedock.adoc]: Hi,

I am wondering if we shouldn't include inline version of devfile directly in the documentation.

The current link points to my personal repository, which is not guaranteed to stay as-is forever.

File: modules/administration-guide/pages/running-containers-with-kubedock.adoc [hunk 5]
```diff
@@ -0,0 +1,66 @@
+:_content-type: ASSEMBLY
+:description: Running containers with kubedock
+:keywords: kubedock, container
+:navtitle: Running containers with kubedock
+:page-aliases:
+
+[id="running-containers-with-kubedock"]
+= Running containers with kubedock
+
+link:https://github.com/joyrex2001/[Kubedock] is a minimal container engine implementation that gives you a Podman-/docker-like experience inside a {prod-short} workspace. Kubedock is especially useful when dealing with ad-hoc, ephemeral, and testing containers, such as in the use cases listed below:
```
Comment (agiertli) [inline modules/administration-guide/pages/running-containers-with-kubedock.adoc]: The correct link is this:
https://github.com/joyrex2001/kubedock

File: modules/administration-guide/pages/running-containers-with-kubedock.adoc (hunk 4 above)
Comment (deerskindoll) [reply modules/administration-guide/pages/running-containers-with-kubedock.adoc]: great point, will do

File: modules/administration-guide/pages/running-containers-with-kubedock.adoc (hunk 5 above)
Comment (deerskindoll) [reply modules/administration-guide/pages/running-containers-with-kubedock.adoc]: fixed

File: modules/administration-guide/pages/running-containers-with-kubedock.adoc [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+* The `podman generate kube` command is not supported.
+* `--env` option causes the `podman run` command to fail.
+====
+
+.Prerequisites
+* An image compliant with link:https://docs.openshift.com/container-platform/{ocp4-ver}/openshift_images/create-images.html#images-create-guide-openshift_create-images[Openshift Container Platform guidelines].
+
+.Process
+* Add `KUBEDOCK_ENABLED=true` environment variable to the devfile.
+** (OPTIONAL) Use the `KUBEDOCK_PARAM` variable to specify additional kubedock parameters.
```
Comment (agiertli) [inline modules/administration-guide/pages/running-containers-with-kubedock.adoc]: User can see available kubedock params here:
https://github.com/joyrex2001/kubedock/blob/master/cmd/server.go

Or by executing:
``` 
kubedock server --help
```

File: modules/administration-guide/pages/running-containers-with-kubedock.adoc [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+[source,yaml,subs="+quotes,attributes"]
+----
+schemaVersion: 2.2.0
+metadata:
+  name: superheroes-workshop
+components:
+  - name: tools
+    container:
+      image: quay.io/devfile/universal-developer-image:latest
+      #image: quay.io/openshiftlabs/quarkus-workshop-stack:3.9
```
Comment (agiertli) [inline modules/administration-guide/pages/running-containers-with-kubedock.adoc]: Please remove the #image line.

File: modules/administration-guide/pages/running-containers-with-kubedock.adoc [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+----
+# kubedock server --help
+----
+
+.Example
+[source,yaml,subs="+quotes,attributes"]
+----
+schemaVersion: 2.2.0
+metadata:
+  name: superheroes-workshop
```
Comment (agiertli) [inline modules/administration-guide/pages/running-containers-with-kubedock.adoc]: Perhaps rename to:
name: kubedock-sample-devfile

**Comment by deerskindoll** [comment]
@agiertli I'll have add the doc to admin guide navigation. do you have any preferences regarding where to place the doc?

**Comment by agiertli** [comment]
> @agiertli I'll have add the doc to admin guide navigation. do you have any preferences regarding where to place the doc?

Hi, this is a user guide content, not admin. Perhaps under Getting Started https://eclipse.dev/che/docs/stable/end-user-guide/getting-started-with-che/ or a brand new section.

**Comment by deerskindoll** [comment]
@agiertli I'll have add the doc to admin guide navigation. do you have any preferences regarding where to place the doc?

> > @agiertli I'll have add the doc to admin guide navigation. do you have any preferences regarding where to place the doc?
> 
> Hi, this is a user guide content, not admin. Perhaps under Getting Started https://eclipse.dev/che/docs/stable/end-user-guide/getting-started-with-che/ or a brand new section.

My mistake. Got it!

**Review by ibuziuk — APPROVED** [review]

**Pull request merged**